### ISSUE TITLE
Reproduce ResNet Result

### ISSUE THREAD

**Issue by chihhuiho** [zhiyuanyou/UniAD#22]
Hello~ Thanks for the nice work and code. I was trying to reproduce the (image/pixel) auroc for ResNet50 and ResNet101, which are (92.4/96) and (92.2/95.9) in the paper, respectively. However, with the provided config file in MVTec folder and the suggested layer changes (https://github.com/zhiyuanyou/UniAD/blob/main/experiments/MVTec-AD/config.yaml#L89), the reproduced result for ResNet50 and ResNet101 are (86.6/94.1) and (87.19/94.66). Could you please provide the config file and checkpoint for resnet50 and resnet101? Thank you in advance.

**Comment by zhiyuanyou** [comment]
Hi~ 

I have checked your email and both attached config files. **_Currently, I could not get the original config files and checkpoints, since I have left SenseTime._** 

I want to verify several things. 

1. Did the results of resnet-18 & resnet-34 consistent with the paper? 
2. How many GPUs did you use in your experiments?
3. Since the main difference is about image AUROC, did you interpret the results right (see Sec. 3.1 in README)?

By the way, I think that the image size could be a difference. If other things are all right, setting the image size as 256 could bring better results. 

I will contact some of my previous colleagues to see whether I could find the original config files.

**Comment by chihhuiho** [comment]
Hi~ Great thanks for the quick response. 
1. I have not tried resnet 18 & resnet 34 yet. I will give you an update once I have it.
2. Currently I am using 1 GPU with batch size 8 (torch.distributed.launch). 
3. To my best understanding, I think I am interpreting the result correctly.
4. Sure, I can try different input size. I am using 224x224 for the above result. 

Basically, I just change the efficient in the config to resnet without modifying anything else (https://github.com/zhiyuanyou/UniAD/blob/main/experiments/MVTec-AD/config.yaml#L84).
If you can recall what else modification has been done for resnet, it will be great to share. Thank you again for the kind and quick response.

**Comment by chihhuiho** [comment]
Hi ~,
I tried resnet18 and resnet34, and I can reproduce the result. Basically, I just change the model to resnet18/resnet34 using the config file that I sent you, without changing any hyperparameters. Is there any possible reason that the config file works for resnet18/resnet34, but not for resnet50/resnet101?  In addition, changing the input size to 256 also boosts the performance a little. 
Thanks for the pointer

**Comment by zhiyuanyou** [comment]
Actually, I did not do any special design for different resnets. 
The resnet18 & resnet34 use **_BasicBlock_**, while the resnet50 & resnet101 use **_Bottleneck_**. Since the resnet is loaded with `strict=False`, some load errors, _i.e._, caused by keys, could happen. 
You can check this log (https://github.com/zhiyuanyou/UniAD/blob/main/models/backbones/resnet.py#L401) to see if there are some missing keys for resnet50 & resnet101.

**Comment by zhiyuanyou** [comment]
Another difference is the number of GPUs. I used 8 GPUs for training, which means `batch_size=64`. How about increasing the batch_size ?

**Comment by chihhuiho** [comment]
For ResNet50, I tried the following (1) increasing the batch size to 64 , (2) changing the input size to 256 and (3) change different seed. Unfortunately, none of them really works. The image auroc still stays around 87. I also make sure that the image net pretrained weight is loaded.  

The only difference between the ResNet18/ResNet34 and ResNet50 experiment is by changing this line
https://github.com/zhiyuanyou/UniAD/blob/main/experiments/MVTec-AD/config.yaml#L84 
But somehow ResNet18/ResNet34 works. I will try wideresnet (the one provided in the backbone_info) and see how it works. Thanks. You are the best author that I ever met :)

**Comment by chihhuiho** [comment]
I also observe the training loss of resnet50 much lower than that of resnet34. Not sure whether if this is the reason of resnet50's worse performance. Again same config file, only the backbone model is changed. The below training loss can be simply reproduced by cloning this repo and change the model of L84 in the config file to resnet50.

![image](https://github.com/zhiyuanyou/UniAD/assets/8252287/93d4874a-fd33-4d6b-8d3e-8717f19e715d)

**Comment by zhiyuanyou** [comment]
Because the performance gap is relatively large, the only element that could cause such a difference is the feature layer. So you could have a try with `outlayers: [1,2,3,4], outlayers: [1,2,3], outlayers: [1,2], outlayers: [3,4]`.
